Include more digestible bits for the introduction of the book (#9)
godot-rust/gdext-book: src/intro/hello-world.md

  ~ file, You can obtain one at https://mozilla.org/MPL/2.0/.
-->

# Hello World

This page shows you how to develop your own small extension library and load it from Godot. The tutorial is heavily inspired by
[Creating your first script][tutorial-begin] from the official Godot documentation. It is recommended to follow that alongside this tutorial,
in case you're interested how certain GDScript concepts map to Rust.

## Table of contents
<!-- toc -->

## Connecting Godot and Rust

First things first.  
For the following objective, you are going to need

- A Cargo crate of your own;

> - A Godot project created;

Reviewer: This reads like the user already needs to set these things up -- however you explain in the next 2 sections to do exactly that. 

Rather formulate it in a way that the following sections are going to explain that, not that the user needs it.

Author: ```suggestion
First things first.  
For the following objective, we'll first make

- A Cargo crate of your own;
- A Godot project created;
- ~~me cry~~
```

> You can very much start from an already existing Godot project. Here we'll start from scratch for the sake of completion.

Reviewer: "completeness", not "completion" 🙂

Author: fixed now :3

### Creating the Godot Project

For this example we're going to create each one separated into neighboring folders in the same parent folder. You'll be able to choose locations of

> your own as we progress. Let's start by creating a blank Godot project if you haven't already.

Reviewer: What does "each one" refer to?

Make sure it's completely unambiguous what you refer to.

Author: pretty sure "each of the items" should be clear enough.
```suggestion
For this example we're going to create each of the items separated into neighboring folders in the same parent folder. [...]
```

> ![alt: A screenshot of the GUI of the Godot project manager about to create a new project][img-godot-create-project]

Reviewer: No need for `alt: ` prefix.

Author: not sure how to format this, this often becomes indistinguishable from normal text whenever the image fails to load, becomes distracting to read.

Not much to do here yet, we'll come back to it later.

### Creating the Rust Cargo crate

> To create a Cargo crate, open have your terminal open, navigate to your desired folder and then call `cargo` to create it by typing

Reviewer: Twice "open".

Also end with a colon.

Author: fixed.

```bash
cargo new "your_crate_name" --lib
```

The argument `--lib` will tell `cargo` to create the package from a Library template. From there, we're going to configure it in the file which stores
its metadata, `Cargo.toml`. This file stores a lot of the information we know of what the library package is about and what it uses in order to make

> it work.

Reviewer: > This file stores a lot of the information we know of what the library package is about and what it uses in order to make it work.

This sentence is overly complex and needs to be re-read multiple times to be understood. Phrases like "we know of", "what it is about", "what it uses to make it work" etc. do not really add anything but rather obfuscate the meaning.

I would omit the whole sentence -- in general, this is not the place to explain what `Cargo.toml` is. Users are expected to have a basic understanding of Rust or at least read the Rust book in parallel.

Author: better keep it for completeness. this is what this whole section of the file is about.
```suggestion
This file stores the information we know of the library package and what it uses in order to work with it.
```

Open `Cargo.toml` in your favorite text editor and edit it as the following:

```toml
[package]

> name = "rust_project" # this name will affect the name of the library file

Reviewer: What is the "library file"? Be precise, otherwise it's mistaken with the `.gdextension` file.

```suggestion
name = "rust_project" # Will appear in the filename of the dynamic Rust library you compile.
```

Author: makes sense. will pick it up.

version = "0.1.0" # version of your library (your criteria)
edition = "2021" # Rust edition version (remains unchanged)

[lib]

> crate-type = ["cdylib"] # defines the kind of library it will become

Reviewer: ```suggestion
crate-type = ["cdylib"] # Compile this crate to a dynamic C library.
```

Author: this was intended to describe the keys, not the values.

> [dependencies] # lists what your own library needs in order to function

Reviewer: This is obvious, no comment needed here.

godot = { git = "https://github.com/godot-rust/gdext", branch = "master" }
```

#### Why "`cdylib`"?
<!-- editors note: sole reason this header's here is because this marks a quick spot to go back to later on,

> specially if a new reader is just scrolling by around the book. -->

Reviewer: Makes sense; this could actually be an admonish note.

I would also recommend to give titles to individual notes:
```md
```admonish note title="cdylib"
Text here...
```

Also applies to other notes.

Author: couldn't fit a pair of quotation marks and another of crases (this thing here ` idk the english name of) within the title. it was another point where i just gave up rather than compromising on writing somethine else entirely. ("why cdylib?" vs. just "cdylib")

The `cdylib` crate type is a bit unusual in Rust. Instead of building a binary (`bin`, application) or a library to be utilized by other Rust code

> (`lib`), we create a _dynamic_ library, exposing an interface in the C programming language This dynamic library is loaded by Godot at runtime,

Reviewer: Missing full stop after "programming language".

Author: fixed.

`librust_project.so` is created if you're on Linux:

```log
$ cargo build
   Compiling godot4-prebuilt v0.0.0 (https://github.com/godot-rust/godot4-prebuilt?branch=4.1.1#fca6897d)
   Compiling proc-macro2 v1.0.69
   [...]
   Compiling godot v0.1.0 (https://github.com/godot-rust/gdext?branch=master#66df8f47)
   Compiling rust_project v0.1.0 (/home/wilker/Documentos/Projects/Programming/Godot/Tutorial/rust_project)
    Finished dev [unoptimized + debuginfo] target(s) in 1m 46s
╭───┬─────────────────────────────────┬──────┬─────────┬────────────────╮
│ # │              name               │ type │  size   │    modified    │
├───┼─────────────────────────────────┼──────┼─────────┼────────────────┤
│ 0 │ target/debug/build              │ dir  │   648 B │ 2 minutes ago  │
│ 1 │ target/debug/deps               │ dir  │  3.8 KB │ now            │
│ 2 │ target/debug/examples           │ dir  │     0 B │ 2 minutes ago  │
│ 3 │ target/debug/incremental        │ dir  │    52 B │ 16 seconds ago │
│ 4 │ target/debug/librust_project.d  │ file │   190 B │ now            │
│ ⮞ │ target/debug/librust_project.so │ file │ 62.6 MB │ now            │

> ╰───┴─────────────────────────────────┴──────┴─────────┴────────────────╯

Reviewer: Where is this from? As far as I know, it doesn't appear automatically upon `cargo build`.

Author: `ls` command output from the Nu Shell, included for visual clarity. added the extra line to avoid confusion.

```

### `.gdextension` file

> The job of this file is to provide Godot with pointers to a library that is compiled elsewhere. We're going to use it to connect Godot to your
> compiled library so it can be used within the context of your Godot project.

Reviewer: ```suggestion
This file tells Godot how to load your compiled Rust extension: the path to the dynamic library file, and the 
entry point (function) to initialize it with.
```

Author: the phrasing feels easier to grasp as it is now. explanation bringing up the terms already shows up later on.

your project**. You can learn more about Godot's resource paths [here][ref-resource-paths].
```

#### Other possible target paths

The following is a list of other possible paths depending on the target of your library. You can copy them as-is for this example, replacing
`{myCrate}` with the name of your actual library:

```ini
[configuration] # same as before
entry_symbol = "gdext_rust_init"
compatibility_minimum = 4.1

[libraries]
linux.debug.x86_64 =     "res://../{myCrate}/target/debug/lib{myCrate}.so"
linux.release.x86_64 =   "res://../{myCrate}/target/release/lib{myCrate}.so"
windows.debug.x86_64 =   "res://../{myCrate}/target/debug/{myCrate}.dll"
windows.release.x86_64 = "res://../{myCrate}/target/release/{myCrate}.dll"
macos.debug =            "res://../{myCrate}/target/debug/lib{myCrate}.dylib"
macos.release =          "res://../{myCrate}/target/release/lib{myCrate}.dylib"
macos.debug.arm64 =      "res://../{myCrate}/target/debug/lib{myCrate}.dylib"
macos.release.arm64 =    "res://../{myCrate}/target/release/lib{myCrate}.dylib"
```

> The `..` is a pointer to a parent folder of the end path.
> Which means Godot will look for the library around the parent folder of the project for its path.

Reviewer: Can also be simplified:
```suggestion
The `..` refers to a parent folder, i.e. Godot will look in a path relative to the project directory's parent.
```

Author: a more direct representation:
```suggestion
The `..` is a pointer to a parent folder of the end path.
e.g Godot will look in a path relative to this project directory's parent folder.
```

```admonish tip

> You can also employ the use of synbolic links and git modules and then treat those as regular folders and files. Godot reads those just fine too! 

Reviewer: Nice addition!

You probably meant "symbolic" and "git submodules"?

Author: fixed :3

```

```admonish note
If you specify your cargo compilation target via the `--target` flag or a `.cargo/config.toml` file, the rust library will be placed in a path name
that includes target architecture, and the `.gdextension` library paths will need to match. For example, for M1 Macs 
(`macos.debug.arm64` and `macos.release.arm64`), the path would be `"res://../rust/target/aarch64-apple-darwin/debug/lib{myCrate}.dylib"`.
```

### `extension_list.cfg`

A second file `res://.godot/extension_list.cfg` should be generated once you open the Godot editor for the first time. If not, you can also manually
create it, simply containing the Godot path to your `.gdextension` file:

```text
res://HelloWorld.gdextension
```

---

> From here on out, your Rust project files should be connected to Godot, so now we'll make it actually function.

Reviewer: The line is not needed.

"actually function" is again imprecise. What is left?
It's also fine to move this to the next section without announcing it beforehand.

Author: i admit i just wanted to humor a bit with this line. see also: https://www.youtube.com/watch?v=-A5B6jqtt-c

> ## Making Rust actually function

Reviewer: Describe _what_ is being set up here.

Author: no fun allowed :c

but fair.

but also this feels like it could use an open-ended title like it.

```

---

From here on out, your Rust project files should be connected to Godot, so now we'll make it actually function.

## Making Rust actually function

### Rust entry point

> [Like mentioned earlier](#why-cdylib), our compiled C library needs to expose an _entry point_ to Godot: a C function that can be called through the GDExtension.

Author: the linting here screams at me for using `a code mark` in a header. no idea how to fix it.

There are multiple things going on here:

1. Place the [`prelude`][api-prelude] module from the [`godot`][api-godot] crate into scope.
   This module contains the most common symbols in the gdext API.
2. Define a struct called `MyExtension`. This is just a type tag without data or methods, you can name it however you like.

> 3. Implement the [`ExtensionLibrary`][api-extensionlibrary] trait for our type, and mark it with the `#[gdextension]` attribute.

Reviewer: Thanks a lot for these links! 👍

Author: np :3

The last point declares the actual GDExtension entry point, and the proc-macro attribute takes care of the low-level details.

> #### "Does it load though?"

Reviewer: ```suggestion
### Loading the extension
```

(Level 3 should be OK)

Author: this phrasing implies that loading the extension requires that `on_level_init` is overloaded. i wanted to clarify the intent by having it as a minor section instead

> If you want a quick look at having it load, you can start out by overloading the `on_level_init(InitLevel)` from the
> [`ExtensionLibrary`][api-extensionlibrary] trait like so:

Reviewer: ```suggestion
If you want to see Godot loading your extension, you can override `on_level_init(InitLevel)` from the
[`ExtensionLibrary`][api-extensionlibrary] trait like so:
```

Author: done.

```rust
use godot::prelude::*;

struct MyExtension;

#[gdextension]
unsafe impl ExtensionLibrary for MyExtension {
    fn on_level_init(level: InitLevel) {
        godot_print!("Hello Library!");

        godot_print!("Initialization level: {:?}", level);
    }
}
```

> And then after `cargo build`, Godot should display the messages when you run Godot from the command line

Reviewer: ```suggestion
After `cargo build`, Godot should display the following messages when running it from the command line:
```

Author: meant to write "these", less verbose. still picking your suggestion though.

```suggestion
After `cargo build`, Godot should display these messages when you run Godot from the command line
```

```log
Godot/Tutorial/GodotProject
$ godot4 -e
Initialize GDExtension API for Rust: Godot Engine v4.2.beta1.official
Hello Library!
Initialization level: Core
Godot Engine v4.2.beta1.official.b1371806a - https://godotengine.org
Hello Library!
Initialization level: Servers
Vulkan API 1.3.246 - Forward+ - Using Vulkan Device #0: AMD - AMD Radeon Vega 8 Graphics (RADV RAVEN)

Hello Library!
Initialization level: Scene
Hello Library!
Initialization level: Editor
```

```admonish note
These messages will most likely not show up unless you're either launching [Godot editor from your command line][godot-command-line] by using
`godot -e` like in this example, or making your gdextension file visible for the first time in Godot while opened.

> ```

Reviewer: Instead of this note, we should probably explain the role of the `.godot/extension_list.cfg` file. Once that is there, this is not a problem and there's no probability involved anymore.

Author: i'm learning to use this whole project as i'm writing this, and the extension_list file is actually one of the things i don't know how to explain yet ^^' this bit was an observation from asking questions in the Discord (and trying stuff on my own).

there is a point up in this file briefly mentioning it exists. maybe an explanation there would be fitting?

i was actually considering placing this whole section inside a collapsed version of the Admonish Example block since this is a section that i could see people ignoring without missing much. by "considering" i mean i failed to group it for an hour before i gave up.

Let's break this down.

> 1. We define two fields `speed` and `angular_speed` for the logic. These are regular Rust fields, no magic involved. More about their use later.

Reviewer: The previous order was better, since it followed the order of different symbols appearing in the code.

I appreciate your improvements though (links + formulation), so would just need swapping some numbers 🙂

Author: sure ^^

2. The gdext prelude contains the most common symbols. Less frequent classes are located in the [`engine`][api-engine] module.

3. The `#[derive]` attribute registers `Player` as a class in the Godot engine.
   See [API docs][api-derive-godotclass] for details about `#[derive(GodotClass)]`.

   ```admonish info
   `#[derive(GodotClass)]` _automatically_ registers the class -- you don't need an explicit 
   `add_class()` registration call, or a `.gdns` file as it was the case with GDNative.
   
   You will, however, need to restart the Godot editor for it to take effect.
   ```

4. The optional `#[class]` attribute configures how the class is registered. In this case, we specify that `Player` inherits Godot's
   `Sprite2D` class. If you don't specify the `base` key, the base class will implicitly be `RefCounted`, just as if you omitted the
   `extends` keyword in GDScript.

5. The `#[base]` attribute declares the `sprite` field, which allows `self` to access the base instance (via composition, as Rust does not have
   native inheritance).

   - The field must have type `Base<T>`.

>      - The type `T` considered valid is affected by what class you've chosen in `#[class(base=Sprite2D)]`

Reviewer: Good addition! I would maybe write it a bit more concretely:
```suggestion
     - `T` must match the declared base class, e.g. `Sprite2D` for `#[class(base=Sprite2D)]`.
```

Author: also included "becomes `Base<Sprite2D>`" just in case.
```suggestion
     - `T` must match the declared base class, e.g. `Sprite2D` for `#[class(base=Sprite2D)]` becomes `Base<Sprite2D>`.
```

That's it for the _Hello World_ tutorial! The following chapters will go into more detail about the various features that gdext provides.

> [godot-command-line]: https://docs.godotengine.org/en/stable/tutorials/editor/command_line_tutorial.html

Author: linting marks first and last links to be unused when they're clearly being used, not sure how to fix this one either.

[api-derive-godotclass]: https://godot-rust.github.io/docs/gdext/master/godot/prelude/derive.GodotClass.html
[api-engine]: https://godot-rust.github.io/docs/gdext/master/godot/engine/index.html
[api-extensionlibrary]: https://godot-rust.github.io/docs/gdext/master/godot/prelude/trait.ExtensionLibrary.html
[api-godot]: https://godot-rust.github.io/docs/gdext/master/godot/index.html
[api-sprite2d]: https://godot-rust.github.io/docs/gdext/master/godot/engine/struct.Sprite2D.html
[api-prelude]: https://godot-rust.github.io/docs/gdext/master/godot/prelude/index.html
[ref-resource-paths]: https://docs.godotengine.org/en/stable/tutorials/scripting/resources.html#external-vs-built-in
[tutorial-begin]: https://docs.godotengine.org/en/stable/getting_started/step_by_step/scripting_first_script.html
[tutorial-full-script]: https://docs.godotengine.org/en/stable/getting_started/step_by_step/scripting_first_script.html#complete-script
[issue-no-reload]: https://github.com/godotengine/godot/issues/66231
[img-godot-create-project]: /resources/images/godot_project_creation.png
"A screenshot of the GUI of the Godot project manager about to create a new project"

> [^ffi]: FFI: [Foreign Function Interface](https://en.wikipedia.org/wiki/Foreign_function_interface)

Reviewer: This could be a regular link, no need for a footnote.

Author: a quick scroll down for a term that's never used again in this page, rather than an external link. but that's fair. change made.

found out alt text also works in links. cool.
